### ISSUE TITLE
Many KeyErrors reported in Tribler 7 alpha 3

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#2765]
I'm receiving many, many reports about a `KeyError` in the (hidden) tunnel community. I think this issue should be prioritised.

Examples:

```
Traceback (most recent call last):
File "TriblerGUI\event_request_manager.py", line 91, in on_read_data
RuntimeError: [Failure instance: Traceback: <type 'exceptions.KeyError'>: 'LibNaCLPK:\x9bg\x84\xfa\xb64;\xf8\x85\xf9\x1c\xfa\x8fb\x81\xc1P\t\xfeW\xde\xc3\x03\x1ct\xf9\xf55\x1e&\x16\x00Id\x9f\x01L\xcc\xc1\x11;\xc4\xcfa\xa2\x8fa>\x00%\xc4\r%x$\xd0)\xe5(i\xb42eR'
lib\site-packages\twisted\internet\base.py:1195:run
lib\site-packages\twisted\internet\base.py:1204:mainLoop
lib\site-packages\twisted\internet\base.py:825:runUntilCurrent
lib\site-packages\twisted\internet\task.py:239:__call__
--- <exception caught here> ---
lib\site-packages\twisted\internet\defer.py:149:maybeDeferred
Tribler\dispersy\util.py:37:helper
Tribler\community\tunnel\tunnel_community.py:441:do_circuits
Tribler\community\tunnel\tunnel_community.py:538:create_circuit
]
```

```
Traceback (most recent call last):
File "TriblerGUI\event_request_manager.py", line 91, in on_read_data
RuntimeError: Unhandled Error
Traceback (most recent call last):
File "lib\multiprocessing\process.py", line 114, in run

File "TriblerGUI\core_manager.py", line 61, in start_tribler_core

File "lib\site-packages\twisted\internet\base.py", line 1195, in run

File "lib\site-packages\twisted\internet\base.py", line 1204, in mainLoop

--- <exception caught here> ---
File "lib\site-packages\twisted\internet\base.py", line 798, in runUntilCurrent

File "Tribler\community\tunnel\hidden_community.py", line 250, in monitor_downloads

File "Tribler\community\tunnel\hidden_community.py", line 640, in create_introduction_point

File "Tribler\community\tunnel\tunnel_community.py", line 538, in create_circuit

exceptions.KeyError: 'LibNaCLPK:\xd7P\x8er2o\xeb?fB\xc6U)\xd4i`\xab\xf5\x0f\xcd\x02r1\xef\xf0\xee\x83g+\x01\xe7T\x94V\x06\xcfb\xa8{r\xc1}\xb7#\xb0uX\x053\x196 \xee\xf0\xbaK\x14\xa3U\x8b\x1fJ(m'
```

```
Traceback (most recent call last):
File "TriblerGUI\event_request_manager.py", line 91, in on_read_data
RuntimeError: Unhandled Error
Traceback (most recent call last):
File "lib\multiprocessing\process.py", line 114, in run

File "TriblerGUI\core_manager.py", line 61, in start_tribler_core

File "lib\site-packages\twisted\internet\base.py", line 1195, in run

File "lib\site-packages\twisted\internet\base.py", line 1204, in mainLoop

--- <exception caught here> ---
File "lib\site-packages\twisted\internet\base.py", line 798, in runUntilCurrent

File "Tribler\community\tunnel\hidden_community.py", line 250, in monitor_downloads

File "Tribler\community\tunnel\hidden_community.py", line 640, in create_introduction_point

File "Tribler\community\tunnel\tunnel_community.py", line 538, in create_circuit

exceptions.KeyError: 'LibNaCLPK:\x93\x9b\xd97<\xbe;\xaf2\xa1\'\x05\xc2NU\xae\xc4\xaf\xf8+*cV\x03\xd05$\xc9!qQ\x13\x92t\x9f\x87\xc3\x17K\xe2\x02\xd9Jd\xf6\xa8t\xf2\xe2\x86\xd2:_2\x93\xf1\x01\xb3O5\x87Jq"'
```

@qstokkink if you want a distraction from thesis work and/or the Windows bug, can you take a look at this issue and see whether there is an appropriate fix? Thanks!

**Comment by qstokkink** [comment]
The only way this could happen is if another member is part of  `self.dispersy_yield_verified_candidates()` but hasn't been contacted through an `introduction_request` or an `introduction_response`.

A.f.a.i.k. this shouldn't&trade; be possible. I am not seeing any obvious scenario in which this would happen.

**Comment by devos50** [comment]
@qstokkink in that case, I guess they should have been removed by the `do_remove` method in `tunnel_community.py`. Any idea how this can happen?

**Comment by synctext** [comment]
no idea what this is.. Suggestion, can we add some debug statements/catch it/various asserts on our suspected causes? Then use the power of the crash reporter to detect the cause with our users in the wild.

**Comment by qstokkink** [comment]
@devos50 Well, that's the peculiar thing: `do_remove()` should only remove entries from `self.exit_candidates` if they are not in `self.dispersy_yield_verified_candidates()` anymore. So `self.dispersy_yield_verified_candidates()` _should_ always be a subset of the entries in `self.exit_candidates`.

The fix for this would be easy enough, just add a check [before retrieving the value](https://github.com/Tribler/tribler/blob/devel/Tribler/community/tunnel/tunnel_community.py#L538):
```python
if pubkey not in self.exit_candidates:
    logging.error("%s not found in %s" % (str(pubkey), str(self.exit_candidates.keys())))
    continue
```

I would like to stress that even though this would stop it from crashing, the cause of the actual bug still remains at large. @devos50 can we somehow feed these errors silently to the crash reporter on the back-end (so without _actually_ crashing)?

**Comment by egbertbouman** [comment]
It may be due to a race condition: create_circuits is called on the main thread and  on_introduction_request/on_introduction_response on the reactor thread. I'll let my pc run overnight, and hope I'll get the error...

**Comment by egbertbouman** [comment]
Unfortunately, I couldn't reproduce this error. But it looks like this happens when a candidate's member changes.

One way this can occur is when an introduced candidate (from an introduction-response message) that is already in stumble/walk is merged with a candidate the is not in stumble/walk. This leads to a change of member, without exit_candidates being updated accordingly.

If this is indeed what's happening, ignoring it is already fine. But since the exit_candidates thing needs refactoring anyway, I'll do that as well.

**Comment by qstokkink** [comment]
@egbertbouman Quick idea: could you add artificial delay (`time.sleep()`) to `create_circuit` and/or the intro request-response handlers to draw out the race condition?

**Comment by egbertbouman** [comment]
@qstokkink Unfortunately, this won't work because everything runs one a single thread (i.e., it's not a race condition).

**Comment by synctext** [comment]
any ideas on this issue?

**Comment by devos50** [comment]
Yes, it should be fixed by #2811 but there are some very minor issues with that PR before it can be merged. I will solve these tomorrow myself and release a beta version.

**Comment by synctext** [comment]
thnx, great news
